src/mpid/ch4/ch4_api.txt

```diff
@@ -0,0 +1,532 @@
+     SHM*: origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count, target_datatype, win, winattr, request
+  mpi_win_flush_local : int
+      NM*: rank, win, addr
+     SHM*: rank, win
+  mpi_win_detach : int
+      NM : win, base-2
+     SHM : win, base-2
+  mpi_compare_and_swap : int
+      NM*: origin_addr, compare_addr, result_addr, datatype, target_rank, target_disp, win, addr, winattr
+     SHM*: origin_addr, compare_addr, result_addr, datatype, target_rank, target_disp, win, winattr
+  mpi_raccumulate : int
+      NM*: origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count, target_datatype, op, win, addr, winattr, request
+     SHM*: origin_addr, origin_count, origin_datatype, target_rank, target_disp, target_count, target_datatype, op, win, winattr, request
+  mpi_rget_accumulate : int
+      NM*: origin_addr, origin_count, origin_datatype, result_addr, result_count, result_datatype, target_rank, target_disp, target_count, target_datatype, op, win, addr, winattr, request
+     SHM*: origin_addr, origin_count, origin_datatype, result_addr, result_count, result_datatype, target_rank, target_disp, target_count, target_datatype, op, win, winattr, request
+  mpi_fetch_and_op : int
+      NM*: origin_addr, result_addr, datatype, target_rank, target_disp, op, win, addr, winattr
+     SHM*: origin_addr, result_addr, datatype, target_rank, target_disp, op, win, winattr
+  mpi_win_allocate : int
+      NM : size-2, disp_unit-2, info, comm, baseptr, win-2
+     SHM : size-2, disp_unit-2, info, comm, baseptr, win-2
+  mpi_win_flush : int
+      NM*: rank, win, addr
+     SHM*: rank, win
+  mpi_win_flush_local_all : int
+      NM*: win
+     SHM*: win
+  mpi_win_unlock_all : int
+      NM*: win
+     SHM*: win
+  mpi_win_create_dynamic : int
+      NM : info, comm, win-2
+     SHM : info, comm, win-2
+  mpi_rget : int
+      NM*: origin_addr-2, origin_count, origin_datatype, target_rank, target_disp, target_count, target_datatype, win, addr, winattr, request
+     SHM*: origin_addr-2, origin_count, origin_datatype, target_rank, target_disp, target_count, target_datatype, win, winattr, request
+  mpi_win_sync : int
+      NM*: win
+     SHM*: win
+  mpi_win_flush_all : int
+      NM*: win
+     SHM*: win
+  mpi_get_accumulate : int
+      NM*: origin_addr, origin_count, origin_datatype, result_addr, result_count, result_datatype, target_rank, target_disp, target_count, target_datatype, op, win, addr, winattr
+     SHM*: origin_addr, origin_count, origin_datatype, result_addr, result_count, result_datatype, target_rank, target_disp, target_count, target_datatype, op, win, winattr
+  mpi_win_lock_all : int
+      NM*: assert, win
+     SHM*: assert, win
+  rank_is_local : int
+      NM*: target, comm
+  av_is_local : int
+      NM*: av
+  mpi_barrier : int
+      NM*: comm, errflag
+     SHM*: comm, errflag
+  mpi_bcast : int
+      NM*: buffer, count-3, datatype, root, comm, errflag
```

Comment:
would it better to define these arguments a list of 2-tuples (name,type)? more verbose, but it would eliminate the need for the mapping below which feels somewhat disconnected from the API arguments. the downside would be in allowing inconsistencies to go unnoticed.

Reply:
> the downside would be in allowing inconsistencies to go unnoticed.

I think forcing the maintainer to go extra for deal with these inconsistencies is a big plus.

If we using 2-tuples, we might as well go back to the C declarations, which I always find very difficult to read and scan for subtleties.

Reply:
Note that this `count-3` will get replaced with `MPI_Aint` soon as part of the large count development. In fact, most of the `int/MPI_Aint` variants will get unified to `MPI_Aint` very soon.